module/applications/actor/base-sheet.mjs

```diff
@@ -1159,6 +1134,38 @@ export default class ActorSheet5e extends ActorSheet {
 
   /* -------------------------------------------- */
 
+  /**
+   * Adjust the preparation mode of a dropped spell depending on the drop location on the sheet.
+   * @param {object} itemData    The item data requested for creation. **Will be mutated.**
+   * @param {DragEvent} event    The original drag event.
+   */
+  _onDropSpell(itemData, event) {
+    if ( itemData.type !== "spell" || itemData.system.level == 0 ) return;
+
+    // Determine the section it is dropped on, if any.
+    const mode = event.target.closest("div[data-type='spell']")?.dataset;
+
+    // If the spell is dropped on a section, set the mode to be that type.
+    if ( mode && mode.level > 0 ) itemData.system.preparation.mode = mode["preparation.mode"];
+
+    // If the spell is not dropped on a section, and on an NPC with no caster levels, set the spell to be innate.
+    else if ( this.document.type === "npc" && !this.document.system.details.spellLevel ) {
+      itemData.system.preparation.mode = "innate";
+    }
+
+    // If the spell is dropped on an actor with only pact magic (and other classes having 'none'), set the spell to be pact.
+    else if ( this.document.type === "character" ) {
+      const modes = this.document.itemTypes.class.reduce((acc, cls) => {
+        if ( cls.spellcasting?.type === "pact" ) acc.pact = true;
+        else if ( cls.spellcasting?.type === "leveled" ) acc.leveled = true;
+        return acc;
+      }, {pact: false, leveled: false});
+      if ( modes.pact && !modes.leveled ) itemData.system.preparation.mode = "pact";
```

Comment:
I might be misreading this, but wouldn't it theoretically be more efficient to rewrite this using a `for` loop?
```suggestion
      let pact = false;
      let leveled = false;
      for (const itemClass of this.document.itemTypes.class) {
        pact |= (itemClass.spellcasting?.type === "pact");
        leveled |= (itemClass.spellcasting?.type === "leveled");
        if (pact && leveled) break;
      }
    if (pact && !leveled) itemData.system.preparation.mode = "pact";
```
This could technically be made even more efficient with an if/else if/else statement, but I feel the above is a good balance between efficiency and readability for what is going to likely be a very small array.

Also, what is `itemData.system.preparation.mode` set to if `pact` is false or `leveled` is true (i.e., that last `if` condition is false)?

Reply:
Then no change is made, which is the idea.

I feel like the current `reduce` is much more readable. The only 'gain' from something more efficient would be for abserd actors with 20+ different classes.

Reply:
Rather than iterating over `this.document.itemTypes.class`, which has to first filter over all items on the actor, you can instead use the `this.actor.classes` object which is pre-filtered.

If no spellcasting classes are found on the player, it perhaps might make sense to set to preparation mode to something like `atwill` rather than leave it as `leveled`. This would accommodate non-spellcasters who get certain spells from their race or from a feat.